File: test/analysis/partial_declaration_test.rb
# frozen_string_literal: true

require_relative "../test_helper"
require_relative "../../lib/herb/analysis/partial_declaration"

class PartialDeclarationTest < Minitest::Spec
  FILE = "app/views/posts/_card.html.erb"

  def declaration_for(source)
    document = Herb.parse(source, strict_locals: true).value

    Herb::Analysis::PartialDeclaration.from_document(document, FILE)
  end

  test "reports no declaration for a partial without strict locals" do
    declaration = declaration_for("<div></div>")

    refute declaration.has_declaration
    assert_empty declaration.locals
  end

  test "collects a required local" do
    declaration = declaration_for("<%# locals: (title:) %>\n<h1></h1>")

    assert declaration.has_declaration
    assert_equal ["title"], declaration.required_locals
    assert_empty declaration.optional_locals
  end

  test "collects an optional local" do
    declaration = declaration_for("<%# locals: (title: nil) %>\n<h1></h1>")

    assert_equal ["title"], declaration.optional_locals
    assert_empty declaration.required_locals
  end

  test "records a keyword rest" do
    declaration = declaration_for("<%# locals: (title:, **rest) %>\n<h1></h1>")

    assert declaration.has_keyword_rest
    assert_equal ["title"], declaration.required_locals
  end

  test "records where the declaration is" do
    declaration = declaration_for("<%# locals: (title:) %>\n<h1></h1>")

    assert_equal 1, declaration.location["line"]
  end

  test "accepts any local when the partial declares none" do
    assert declaration_for("<div></div>").accepts?("anything")
  end

  test "accepts any local when the partial takes a keyword rest" do
    assert declaration_for("<%# locals: (title:, **rest) %>").accepts?("anything")
  end

  test "rejects a local the partial does not declare" do
    declaration = declaration_for("<%# locals: (title:) %>")

    assert declaration.accepts?("title")
    refute declaration.accepts?("subtitle")
  end

  test "round trips through a serialized form" do
    declaration = declaration_for("<%# locals: (title:, subtitle: nil) %>")
    restored = Herb::Analysis::PartialDeclaration.from(declaration.to_h)

    assert_equal declaration.to_h, restored.to_h
    assert_equal ["title"], restored.required_locals
  end
end

File: rust/herb-analysis/tests/partial_declaration_test.rs
use herb::herb::{parse_with_options, ParserOptions};
use herb_analysis::partial_declaration::PartialDeclaration;

const FILE: &str = "app/views/posts/_card.html.erb";

fn declaration_for(source: &str) -> PartialDeclaration {
  let options = ParserOptions {
    strict_locals: true,
    ..Default::default()
  };

  let result = parse_with_options(source, &options).expect("parse failed");

  PartialDeclaration::from_document(&result.value, FILE)
}

#[test]
fn reports_no_declaration_for_a_partial_without_strict_locals() {
  let declaration = declaration_for("<div></div>");

  assert!(!declaration.has_declaration);
  assert!(declaration.locals.is_empty());
}

#[test]
fn collects_a_required_local() {
  let declaration = declaration_for("<%# locals: (title:) %>\n<h1></h1>");

  assert!(declaration.has_declaration);
  assert_eq!(declaration.required_locals(), vec!["title"]);
  assert!(declaration.optional_locals().is_empty());
}

#[test]
fn collects_an_optional_local() {
  let declaration = declaration_for("<%# locals: (title: nil) %>\n<h1></h1>");

  assert_eq!(declaration.optional_locals(), vec!["title"]);
  assert!(declaration.required_locals().is_empty());
}

#[test]
fn records_a_keyword_rest() {
  let declaration = declaration_for("<%# locals: (title:, **rest) %>\n<h1></h1>");

  assert!(declaration.has_keyword_rest);
  assert_eq!(declaration.required_locals(), vec!["title"]);
}

#[test]
fn records_where_the_declaration_is() {
  let declaration = declaration_for("<%# locals: (title:) %>\n<h1></h1>");

  assert_eq!(declaration.location.expect("location").line, 1);
}

#[test]
fn accepts_any_local_when_the_partial_declares_none() {
  assert!(declaration_for("<div></div>").accepts("anything"));
}

#[test]
fn accepts_any_local_when_the_partial_takes_a_keyword_rest() {
  assert!(declaration_for("<%# locals: (title:, **rest) %>").accepts("anything"));
}

#[test]
fn rejects_a_local_the_partial_does_not_declare() {
  let declaration = declaration_for("<%# locals: (title:) %>");

  assert!(declaration.accepts("title"));
  assert!(!declaration.accepts("subtitle"));
}
